tools/migration/internal/migration.go

```diff
@@ -30,17 +30,27 @@ type Migration interface {
 	Validate(oldRepo, newRepo *os.File) error
 }
 
+type RepoHelper interface {
+	GetOldRepo() (*os.File, error)
```

Comment:
I'm concerned that this method / general pattern is not providing the right thing for our migrations.  Most migrations will need to do their own thing with a repo and an `*os.File` is too opinionated an abstraction to write migrations simply.  For example, in the first simple migration I've been thinking about converting chain metadata from json to cbor.  To do this the migrating function should open a datastore, read all of a certain kind of key and rewrite the data at that key in the new format.  To do this the migration really needs a path to get a datastore object from `badgerds.NewDatastore`, an os.File *could* work, but the most straightforward thing you would do with it is read the path using `Read` or `Readdir`.

Not saying we need to do exactly what ipfs does, but the prior art here makes a lot of sense to me.  [Migration interfaces](https://github.com/ipfs/fs-repo-migrations/blob/master/go-migrate/migrate.go#L14) are given the power to [read paths](https://github.com/ipfs/fs-repo-migrations/blob/master/go-migrate/cli.go#L12)

Reply:
Correcting myself: the keys should be read from the old datastore and written to the new datastore.  This implies that it is also better to define access to the new repo in terms of a path.

Reply:
Going back to the design doc I think this isn't such a big change to the interface.  I think we just need to change
```golang
// Migrate runs the migration steps on a copy of the repo
func (mig *Migrator) Migrate(newRepo os.File,) error
// Validate returns error if migration tests failed, describing why
func (mig *Migrator) Validate(oldRepo, newRepo os.File) error
```

to 

```
// Migrate runs the migration steps on a copy of the repo
func (mig *Migrator) Migrate(newRepoPath string,) error
// Validate returns error if migration tests failed, describing why
func (mig *Migrator) Validate(oldRepoPath, newRepoPath string) error
```

Reply:
Ok, I guess I misunderstood a previous comment from @anorth  (which I'm too lazy to hunt down) -- I thought we specifically did not want to give individual migrations access to the file system, but control the type of access and hand them file pointers.

Reply:
I vaguely remember that too and I can see us wanting to do something to restrict what the migration can do.  If we want to go down that road we'll need to fine tune which interfaces the migration has access to so that we can restrict but in a way that is readily consumed by the migration.  For example the interface would need to return a collection of datastores, files etc.  This gets pretty messy because the interface might end up needing to become migration dependent.

I would argue that the whole song and dance around copying the entire repo and only giving the path of the copy to the migration gets us both isolation and the needed flexibility.

Reply:
If it only gets the path to the copy, how does it validate the results though?